sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -256,6 +257,22 @@ class FindDataSourceTable(sparkSession: SparkSession) extends Rule[LogicalPlan]
       QualifiedTableName(table.identifier.catalog.get, table.database, table.identifier.table)
     val catalog = sparkSession.sessionState.catalog
     val dsOptions = DataSourceUtils.generateDatasourceOptions(extraOptions, table)
+
+    // Invalidate the cache if the table is already cached but options have changed.
+    val cachedPlan = catalog.getCachedTable(qualifiedTableName)
+    if (cachedPlan != null) {
+      cachedPlan match {
+        case LogicalRelationWithTable(
+        HadoopFsRelation(_, _, _, _, _, options), _) =>
+          val prevOptions = new CaseInsensitiveStringMap(options.asJava)
+          val newOptions = new CaseInsensitiveStringMap(dsOptions.asJava)
+          if (prevOptions != newOptions) {
+            catalog.invalidateCachedTable(qualifiedTableName)
+          }
+        case _ =>
+      }
+    }
+
     catalog.getCachedPlan(qualifiedTableName, () => {
```

Comment:
Let's compare the data source option with the result plan.
If different, construct a plan with the new dsOptions.